Add transforms for shared attributes.

coderkevin opened the pull request:
Fixes: #370 

This adds transforms for the 5 blocks which share attributes:

- woocommerce/product-best-sellers
- woocommerce/product-category
- woocommerce/product-new
- woocommerce/product-on-sale
- woocommerce/product-top-rated

#### Accessibility

No changes to any interfaces, just using existing Gutenberg transforms.

### How to test the changes in this Pull Request:

1. Create a block of one of the types, set columns, set categories
2. Transform it to each of the other types in turn. Each of the 20 transforms possible should be tested from each block to each other block.

coderkevin committed:
Add transforms for shared attributes.

This adds transforms for the 5 blocks which share attributes:

- woocommerce/product-best-sellers
- woocommerce/product-category
- woocommerce/product-new
- woocommerce/product-on-sale
- woocommerce/product-top-rated

justinshreve commented on assets/js/blocks/product-best-sellers/index.js:
```diff
@@ -28,6 +29,20 @@ registerBlockType( 'woocommerce/product-best-sellers', {
+		from: [
+			{
+				type: 'block',
+				blocks: [
```
Should we consider having a helper method that can return a list of blocks, minus the one being transformed to? If more blocks are added, each of these `transforms` will need to be updated.

ryelle replied:
I wonder if we can do this automatically; get the list of blocks in the woocommerce category that have the rows attribute? I think that criteria will get the blocks we want to transform.

Alternately just having one single place where this is hardcoded would be easier to manage.

coderkevin replied:
Given that there is an order to calling `registerBlock` upon load, I don't think we can depend on dynamically querying block types. That list would start out empty for the first `woocommerce/...` block and would only be complete for the last.  I think we will need a static implementation for this.

ryelle commented on assets/js/utils/transforms.js:
```diff
@@ -0,0 +1,24 @@
+ */
+import sharedAttributes from './shared-attributes';
+
+export function makeTransform( toBlock, columns, defaultAttributes ) {
```
This `columns` variable threw me at first since I expected it was related to the attribute `columns` - but I'm not sure you need it? I think you could get rid of `makeSharedAttributesTransform` entirely, [since `createBlock` will only use attributes defined in the block schema.](https://github.com/WordPress/gutenberg/blob/211f65dd854fc1f7d7e54782863ad0629abeff41/packages/blocks/src/api/factory.js#L39)

Actually, I think you can use createBlock directly in the transform, and then you don't even need this helper file. Like, in newest products:

```js
transform: ( attributes ) => createBlock( 'woocommerce/product-new', attributes )
```

ryelle reviewed (changes requested):
Thanks for working on this! It works, but I've got a suggestion for simplifying it further.

coderkevin committed:
Use createBlock directly and rowColumnBlockTypes

This adds rowColumnBlockTypes as an array of woocommerce/* block types
that use row and column layouts

ryelle commented on assets/js/blocks/index.js:
```diff
@@ -0,0 +1,8 @@
+
```
Can you move this file into the utils folder?

ryelle commented on assets/js/blocks/product-best-sellers/index.js:
```diff
@@ -29,6 +31,18 @@ registerBlockType( 'woocommerce/product-best-sellers', {
+		from: [
+			{
+				type: 'block',
+				blocks: difference( rowColumnBlockTypes, [ 'woocommerce/product-best-sellers' ] ),
```
If you use `without` instead of `difference`, this doesn't need to be an array (i also benchmarked it out of curiosity, and it seems `without` is slightly more performant, too)

coderkevin committed:
Move block types into utils, and use without

This uses lodash instead of without instead of difference
It also moves the block types from an index into the shared-attributes
in utils.

ryelle reviewed (approved):
Looks good, transforms are working across blocks, keeping the shared attributes 👍
